specification/dartLangSpec.tex

```diff
@@ -17642,8 +17648,23 @@ \subsection{Type Aliases}
 (\ref{redirectingFactoryConstructors}),
 and it can be used as a superclass, mixin, or superinterface
 (\ref{superclasses}, \ref{mixinApplication}, \ref{superinterfaces}).
-A compile-time error occurs at each of these usages,
-if $F$ does expand to a type variable.
+
+\LMHash{}%
+In each of these cases,
+a compile-time error occurs if $F$ expands to a type variable,
```

Comment:
'Expands to' may be a debatable terminology. I'm using it because I'm thinking about the type alias semantics as a macro-like mechanism: It is defined in terms of substitution, and it occurs at compile time. Do you think about type aliases in a different way, do you have a better term in mind?

The 'expands to' property is associated with the type alias declaration as such, not any particular reference to it (be it `$F$` or `$F$<...>`). So it's a characterization of the substitution as such. This also implies that it is a compile-time concept.

> we have to be explicit that we check this statically by expanding the clause `F<T>`

I think it's better to talk about structural properties of the declaration, and omit all references to an actual substitution. It is true that we want to say something about substitutions in the end, but it's much simpler if we can do all of that in terms of structural properties of the declarations. (In other words, we won't ever consider a rewriting system, the potentially infinite set of terms that we can obtain from an initial term by rewriting, or the process properties of the rewrite system like termination etc., we just look at the finite set of declarations as written).

The 'expands to a type variable' property is described in terms of the _existence_ of a `<typeName>` in the body or in a type parameter bound which is a reference to some other type alias declaration. This is used to define a dependency relation between type alias declarations, and then we just check that graph for cycles.

The _consequence_ is that if `X` is a type variable and we consider the type denoted by `F<X>`, performing the substitutions recursively until the resulting denoted type does not contain any references to any type aliases, it is guaranteed that it won't be `X`. But we don't even need to say that explicitly.

(It would probably be a friendly move to add some more `\rationale{}` text about this. I'll do that tomorrow morning.)